src/script/lua_api/l_object.cpp

```diff
@@ -1410,6 +1410,142 @@ int ObjectRef::l_get_player_control_bits(lua_State *L)
 	return 1;
 }
 
+// Reads in a physics modifier from a table, defaulting to existing values.
+// Returns success.
+static bool read_physics_modifier(lua_State *L, int index,
+		PlayerSAO::PhysicsModifier &modifier)
+{
+	if (!lua_istable(L, index))
+		return false;
+
+	auto type = getstringfield_default(L, index, "type", "multiply");
+	if (type == "add") {
+		modifier.is_add = true;
+	} else if (type != "multiply") {
+		warningstream << "Unknown modifier type " << type << ", defaulting to multiply." << std::endl;
+		warningstream << script_get_backtrace(L);
+	}
+
+	float def = modifier.is_add ? 0.f : 1.f;
+
+	modifier.speed = getfloatfield_default(L, index, "speed", def);
+	modifier.jump = getfloatfield_default(L, index, "jump", def);
+	modifier.gravity = getfloatfield_default(L, index, "gravity", def);
+
+	return true;
+}
+
+// get_effective_physics(self)
+int ObjectRef::l_get_effective_physics(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	PlayerSAO *playersao = (PlayerSAO *)getobject(ref);
+	if (playersao == nullptr)
+		return 0;
+
+	const auto &physics = playersao->getEffectivePhysics();
+
+	lua_newtable(L);
+	lua_pushnumber(L, physics.speed);
+	lua_setfield(L, -2, "speed");
+	lua_pushnumber(L, physics.jump);
+	lua_setfield(L, -2, "jump");
+	lua_pushnumber(L, physics.gravity);
+	lua_setfield(L, -2, "gravity");
+	return 1;
+}
+
+// set_physics_modifier(self, key, modifier_table)
+int ObjectRef::l_set_physics_modifier(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	PlayerSAO *playersao = (PlayerSAO *) getobject(ref);
+	if (playersao == nullptr)
+		return 0;
+
+	const std::string key = readParam<std::string>(L, 2);
+
+	PlayerSAO::PhysicsModifier modifier;
+	if (!read_physics_modifier(L, 3, modifier))
+		throw LuaError("set_physics_modifier expects (string, table)");
+
+	playersao->setPhysicsModifier(key, modifier);
+
+	if (playersao->m_physics_override_set) {
+		warningstream << "set_physics_modifier will have no effect because ";
+		warningstream << "set_physics_override was previously called ";
+		script_log_short_src(L, warningstream);
+		warningstream << std::endl;
+	}
+
+	return 0;
+}
+
+// remove_physics_modifier(self, key)
+int ObjectRef::l_remove_physics_modifier(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	PlayerSAO *playersao = (PlayerSAO *) getobject(ref);
+	if (playersao == nullptr)
+		return 0;
+
+	const std::string key = readParam<std::string>(L, 2);
+
+	playersao->deletePhysicsModifier(key);
+
+	if (playersao->m_physics_override_set) {
+		warningstream << "remove_physics_modifier will have no effect because ";
+		warningstream << "set_physics_override was previously called ";
+		script_log_short_src(L, warningstream);
+		warningstream << std::endl;
```

Comment:
Question here again: Does it make sense to log this to warningstream because
* the mod calling `(set|remove)_physics_modifier` is not at fault
* a warning will already have been logged for the mod that called `set_physics_override`

?

Reply:
Good point

Reply:
Actually, this isn't necessarily logged if the overrides are set before the first modifier.
The point of the warning here is to help debug something that would otherwise be quite confusing

Reply:
It can stay then.